### DISCUSSION TITLE
1.0.0 Trial Use Review Ballot (November 2024)

### DISCUSSION THREAD

**Discussion by ahwagner** [ga4gh/cat-vrs#86] [Announcements]
# Welcome to the Cat-VRS Nov '24 Trial Use review
> Dear reviewer,
> 
> If you are able and willing to participate in the development of these standards, please list individual concerns as separate comments below. All parties are welcome. 
>
>**Some guidance on creating a comment**
>Ideally, you should place a Heading tag on the first line of your comment with a summary heading of the specific issue or question (see first comment below as a model). There is no limit to how many comments you can create. It would also be helpful to first look for a similar or related comment and work within the reply thread of that comment if possible. 
> 
> Thank you for your participation,
> The GKS product leads and co-leads

## Dates to Remember!
- The ballot review period will begin on **Thursday, Nov 28, 2024** and close on **Friday, Dec 6, 2024**.
- There will be an opportunity for a live feedback and review of the process to be held during the [*Categorical Variant Meeting* on Wednesday, Dec 4, 2024](https://www.google.com/calendar/event?eid=NzJ0OGRxN2k5OGp2Nm0xbmg1NmI4NjUybmJfMjAyNDEyMDRUMjAwMDAwWiBpbmZvQGdhNGdoLm9yZw). All VA and Cat-VRS reviewers are welcome to attend.

## Implementer Survey (<5mins)
- For all current, future or possible implementers we would like you to also answer this very short survey to indicate which of the *Trial Use* classes you are most interested in. This will help validate the number of implementers as well as help us keep you informed if future changes occur to your classes of interest. Here are the Implementer's Trial Use Review forms for the [VA-Spec](https://forms.gle/1jP5PrUKpTR4To118) and [Cat-VRS](https://forms.gle/5dVuiErGbv23d1PH6) products, please participate, your contributions help advance the standards and make the process work.

# The List of Classes under Consideration
This ballot is a GKS community trial use review for the following classes:

## Cat-VRS Base classes

- Categorical Variant [[docs](https://cat-vrs.readthedocs.io/en/1.0.0-ballot.2024-11/concepts/catvrs_model.html#categorical-variant)] [[schema](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11.1/schema/cat-vrs/json/CategoricalVariant)] [[example](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/examples/describedVariant-ex1.yaml)]
- Defining Allele Constraint [[docs](https://cat-vrs.readthedocs.io/en/1.0.0-ballot.2024-11/concepts/catvrs_model.html#definingalleleconstraint)] [[schema](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11.1/schema/cat-vrs/json/DefiningAlleleConstraint)] [[example](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/examples/canonicalAllele-ex1.yaml)]
- Defining Location Constraint [[docs](https://cat-vrs.readthedocs.io/en/1.0.0-ballot.2024-11/concepts/catvrs_model.html#defininglocationconstraint)] [[schema](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11.1/schema/cat-vrs/json/DefiningLocationConstraint)] [[example](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/examples/categoricalCnv-ex1.yaml)]
- Copy Number Count Constraint [[docs](https://cat-vrs.readthedocs.io/en/1.0.0-ballot.2024-11/concepts/catvrs_model.html#copycountconstraint)] [[schema](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/schema/cat-vrs/json/CopyCountConstraint)] [[example](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/examples/categoricalCnv-ex1.yaml)]

## Cat-VRS Recipes

- Canonical Allele [[docs](https://cat-vrs.readthedocs.io/en/1.0.0-ballot.2024-11/concepts/recipes.html#canonicalallele)] [[schema](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11.1/schema/cat-vrs/json/CanonicalAllele)] [[example](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/examples/canonicalAllele-ex1.yaml)]
- Protein Sequence Consequence [[docs](https://cat-vrs.readthedocs.io/en/1.0.0-ballot.2024-11/concepts/recipes.html#proteinsequenceconsequence)] [[schema](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11.1/schema/cat-vrs/json/ProteinSequenceConsequence)] [[example](https://github.com/ga4gh/cat-vrs/blob/1.0.0-ballot.2024-11/examples/proteinSequenceConsequence-ex1.yaml)]

Join the discussion on this page!

**Changelog Since September Connect**: https://github.com/ga4gh/cat-vrs/compare/1.0.0-connect.2024-09.1...1.0.0-ballot.2024-11.1

<hr /><em>This discussion was created from the release <a href='https://github.com/ga4gh/cat-vrs/releases/tag/1.0.0-ballot.2024-11.1'>1.0.0 Trial Use Review Ballot (November 2024)</a>.</em>

**Comment by larrybabb** [comment]
### `CategoricalCnv` and `CopyNumberChangeConstraint` SHOULD be added to the TU review...

Currently, the examples only consider the use of `DefiningLocationConstraint` and `CopyNumberCountConstraint` in the context of a `CategoricalCnv` recipe.  If `CategoricalCnv` is NOT up for *trial use* consideration then those constraints should be removed.  If it is then the `CopyNumberChangeConstraint` SHOULD also be added to the TU review since it is part of the `CategoricalCnv` recipe.

**Reply by DanielPuthawala** [reply]
@ahwagner Can you weigh in on this? Should the `CategoricalCNV` recipe be on the ballot?

**Reply by ahwagner** [reply]
IMO the way it is set up now is appropriate for where community implementations are at. Recipes reflect community conventions for certain types of categorical variants, e.g. canonical alleles (NCBI, ClinGen). AFAIK there is no community norm for a categorical CNV. However, there is clearly a need for copy number constraints in ClinVar and some of the VICC KBs.

**Reply by brendanreardon** [reply]
Adding in that moalmanac, and likely CIViC, is interested in using Categorical CNVs with the GeneContextConstraint instead of DefiningLocationConstraint. The GeneContextConstraint is something that we can resolve in the 1.1 release though, it was a bit sidelined in the [relevant discussion thread](https://github.com/ga4gh/cat-vrs/discussions/59#discussioncomment-11103809). 

Also echoing that it seems to me to be okay to hold off on adding `CategoricalCNV` and `CopyNumberChangeConstraint` until a later release. In the meantime instead, we should add an example(s) from NCBI/ClinGen of adding the `CopyNumberCountConstraint` to a CanonicalAllele or ProteinSequenceConsequence (I'm guessing that they are using this w.r.t. CanonicalAllele though). I'll take ownership of drafting up an example that meets this criteria.

**Reply by brendanreardon** [reply]
We moved these comments to https://github.com/ga4gh/cat-vrs/issues/88, as it already related to CNV classes for the next trial use release. Please continue the conversation there! 🙇

**Comment by korikuzma** [comment]
# Add Reference implementation to docs

I've created a Python reference implementation here: https://github.com/ga4gh/cat-vrs-python

I will be updating models soon.

**Reply by brendanreardon** [reply]
Thank you, @korikuzma 🙇! We created an issue to track documentation changes recommended from  this ballot, [#91](https://github.com/ga4gh/cat-vrs/issues/91).

**Reply by brendanreardon** [reply]
I moved this to its own [Issue (#96)](https://github.com/ga4gh/cat-vrs/issues/96), per instruction from Tuesday's meeting. Please continue the discussion there 🙇

**Comment by korikuzma** [comment]
# Add examples

One thing I really like about the new docs for VRS, is that it now includes examples. I think it would be really beneficial to users to add examples for each class.

**Reply by afrubin** [reply]
Thank you for this. The examples are super useful for our project, especially when we are trying to engage with only a subset of the spec (Protein Sequence Consequence) at this stage.

On a related note, the recipe example file is super useful, but it would be great if the `vrs.json` file that is referenced in the `constraints` and `members` sections was also part of the repo for completeness.

**Reply by DanielPuthawala** [reply]
Like a fully-verbose example that drills down all the way to the GKS.core stuff?

**Reply by afrubin** [reply]
I think if one complete example was available it would help implementers, since it would be easier to see what CatVRS requires in one place. Having said that, it might be a lot of work if you don't have those files already so it might be out of scope for this TU ballot.

**Reply by brendanreardon** [reply]
Linking out to this Issue that Daniel created last week to track example updates! [#87](https://github.com/ga4gh/cat-vrs/issues/87)

**Reply by brendanreardon** [reply]
Kori's original comment and subsequent responses are now quoted in the relevant Issue: https://github.com/ga4gh/cat-vrs/issues/87. Please continue to the conversation there 🙇

**Comment by Mrinal-Thomas-Epic** [comment]
# Should `relations` be a mappable concept?
Applies to both `DefiningAlleleConstraint` and `DefiningLocationConstraint`:
* Do we expect different implementations to define their own types of relations? Wouldn't that defeat the purpose of cat-vrs (to be specific and descriptive about variant matching logic)?
* Do we expect implementations to use the other properties of MappableConcept (i.e., label, mappings, extensions)?

**Reply by DanielPuthawala** [reply]
I'm not sure we've really settled that question yet.  AFAIK the current tack is that we can specify some relations, and adopters have those at a minimum, but could define their own custom relations in addition to the base vanilla ones?   @ahwagner and @larrybabb how closely does this reflect your guys' current thinking?

**Reply by DanielPuthawala** [reply]
By way of update, there were three reasons for removing the explicit list of relations from the constraint to Mappable concept instead.  

First, it's more DRY this way, and makes it easier to update relations if they are in one place in the spec instead of being distributed amongst the various constraints.  

Second, it seems to make more sense to specify required relations in the recipes rather than the constraints, wherein any compatibly-typed relations are optional.  Recipes, by contrast, can and do require some minimal relations, e.g. see the Canonical Allele recipe.  

Third, the schema isn't the only place in the spec we can give guidance about which relations to include, and this is especially relevant given that the reference implementation and guidance to implementers are really where relation functions will have to be fleshed out in the first place.  So perhaps guidance should live there?  

I'd be happy to make an issue on this topic, @Mrinal-Thomas-Epic if we want revisit this question, just let me know!

**Reply by brendanreardon** [reply]
We moved this to its own GitHub issue for tracking purposes: https://github.com/ga4gh/cat-vrs/issues/97 😄 Please continue the conversation there! 🙇

**Comment by rhdolin** [comment]
### Categorical Variant class - minor wordsmithing

"Categorical variants exist solely within genomics knowledgebases" - suggest we broaden this, to something like "Categorical variants exist within genomics knowledgebases, as clinical trial eligibility criteria, and within clinical quality measures".

**Reply by brendanreardon** [reply]
Totally valid point. Thank you, Bob! We created an issue ([#91](https://github.com/ga4gh/cat-vrs/issues/91)) to track documentation updates per this ballot, including this thoughtful suggestion.

**Comment by rhdolin** [comment]
### Categorical Variant class - attribute cardinality

1. Consider making categoricalVariant.label required (hard to know what the constraints are constraining otherwise...)
2. categoricalVariant.constraints minimum cardinality should be 1 (why else instantiate the class?)

**Reply by DanielPuthawala** [reply]
1. I'm not sure we want to do this, since `label` is usually more of a human-readibility thing, and I can imagine that lots of CatVars in, say, ClinVar, aren't generated with or are even concerned with being human-readible. and so there `label` would be extraneous.  Aside form that, imo the `properties` of each `constriant` already record what is being constrained already, be it a location, allele, copy number, etc.  That's not to say that you couldn't require `label` to be populated in a recipe specific to your use-case, though!  I'm just not convinced we need it to be required across the board for _all_ use cases.
2. That's not a bad idea.  Will have to noodle on this!

**Reply by brendanreardon** [reply]
As a counter point to requiring minimum cardinality of 1, having the minimum constraints as 0 will allow implementers to pre-allocate and include variant classes that are not yet covered by either cat-vrs or vrs. 

For example, if an implementer had a database that contained canonical alleles and other variant types that are covered by vrs/cat-vrs, but also things like... PD-L1 tumor proportion score, they could use the cat-vrs schema for their entire database, not attach any constraints to biomarkers like PD-L1 TPSs, and use extensions to represent the biomarkers as a placeholder.

**Reply by brendanreardon** [reply]
Hey @rhdolin, just following up on this. We're _not_ going to move this to an Issue to continue the conversation for the time being, but happy to do so in the future if this comes up again.

**Comment by rhdolin** [comment]
### Categorical variant constraints need Boolean operators

categoricalVariant.constraints needs the ability to insert boolean operators (e.g. [constraint A AND (constraint B OR constraint C)]).

**Reply by DanielPuthawala** [reply]
The system certainly does absolutely require boolean operators (including potentially NOT) somewhere in the system, and this is a puzzle that I have lost a lot of sleep over, but I'm currently leaning in the direction that we would probably be best off building that logic into the `recipes`, (and _maybe_ within a constraint itself) but not the baseline `CategoricalVariant` class itself, at least not much.

If a `CategoricalVariant` has constraints A, B, and C specified in `constraints`, then you can effectively think of those constraints as all ANDed together.  A copy-number gain variant in MET CatVar would have stuff specified for the `GeneContextConstraint` (MET) and for the `CopyCountConstraint`, and the members are those contextual variants that are in _MET_ (per the `GeneContextConstriant` AND constitute a CN-Gain (per the `CopyCountConstriant`.

However, I can certainly imagine a recipe you might want to specify for a CatVar where there is, for example, a genomic sequence insertion or deletion, but NOT a change in reading frame.

So one way we could imagine doing this would be to specify that a recipe requires that `DefiningAlleleConstraint` is populated (with appropriate properties), but that the (hypothetical) `ReadingFrameConstraint` is not populated. 

But I think a better way of handling this would be to instead include in `ReadingFrameConstraint`'s properties the option to indicate something like `no change to reading frame`, because this way, (1) we can distinguish between a CatVar that someone has asserted does not impact reading frame from one where noone has made a claim one way or the other, and (2) In think this approach would maybe let us get away without having to have negation (and associated scoping headaches) in the logic of Cat-VRS.  As for AND vs OR logic in recipes, I believe we can already handle these be requiring `allOf` and `anyOf`, respectively.  But I should probably set aside from time in the future to sit down (the more the merrier) and try to prove to myself that those are all true in all relevant cases...

**Reply by rhdolin** [reply]
@DanielPuthawala I hadn't seen allOf or anyOf - can you point me to more information on those? But to the bigger question, I understand the value of recipes, but I think the base boolean syntax needs to be defined in the main standard. Have you considered, for instance, that the data type of CategoricalVariant.constraints be 'expression' rather than just a set of constraints? Seems like we can define or draw upon existing specifications to have a simple syntax of combining constraints with AND, OR, NOT, etc.

**Reply by DanielPuthawala** [reply]
Here's an example, @rhdolin, where the recipe for `CanonicalAllele` requires `allOf` in `constraints`. Granted, this isn't the greatest example, because it's an `AND` requirement over a singleton list, but the logic is still there.

```   CanonicalAllele:
     maturity: trial use
     description: >-
       A canonical allele is defined by an `Allele <https://vrs.ga4gh.org/en/2.x/concepts/MolecularVariation/Allele.html#>`_
       that is representative of a collection of congruent Alleles, each of which depict the same nucleic acid
       change on different underlying reference sequences. Congruent representations of an Allele often exist
       across different genome assemblies and associated cDNA transcript representations.
     allOf:
       - $ref: CategoricalVariant
       - properties:
           constraints:
             minItems: 1
             maxItems: 1
             contains:
               allOf:
                 - $ref: DefiningAlleleConstraint
                 - type: object
                   properties:
                     relations:
                       allOf:
                         - contains:
                             type: object
                             properties:
                               primaryCode:
                                 const: liftover_to
                         - contains:
                             type: object
                             properties:
                               primaryCode:
                                 const: transcribes_to
                       maxContains: 2```

**Reply by DanielPuthawala** [reply]
Here's a better example, though one which isn't yet on the main branch.  [On the Issue 16 branch](https://github.com/ga4gh/cat-vrs/blob/16-update-function-variants/schema/cat-vrs/recipes-source.yaml), the recipe for `FunctionVariant` also uses `allOf`, and this time, for a non-singleton list, requiring that both the `GeneContextConstraint` `AND` `FunctionConstraint` be satisfied.

 ``` FunctionVariant:
    maturity: draft
    description: >-
      A representation of the constraints for matching knowledge about function variants; e.g., gain-of-function or
      loss-of-function.
    allOf:
      - $ref: "/ga4gh/schema/cat-vrs/1.x/json/CategoricalVariant"
      - properties:
          constraints:
            description: >-
              The constraints array must contain exactly two items:
              a GeneContextConstraint and a FunctionConstraint.
            minItems: 2
            maxItems: 2
            contains:
              allOf:
                - $ref: "/ga4gh/schema/cat-vrs/1.x/json/GeneContextConstraint"
                - $ref: "/ga4gh/schema/cat-vrs/1.x/json/FunctionConstraint"
        required:
          - constraints```

**Reply by DanielPuthawala** [reply]
So the boolean logic is in the schemas, just for `Recipes`, and not for the base `CategoricalVariant` class at the moment.   But I'm amenable to changing that if it makes sense.  

One such thing I'm not sure how we'll handle at the moment are compound CatVars, like `EGFR Exon 19 or 21 deletion` or `MET Amplification and TP53 Loss`.   Are those the sort of examples you had in mind?

Also, can you elaborate more on what it means to type CatVar constraints as `expressions`?

**Reply by rhdolin** [reply]
Thanks for this @DanielPuthawala. Still not quite following - where do you define where allOf/anyOf can go? That's kinda what I was getting at with a data type of expression, where the expression is a string that adheres to a defined syntax that formalizes where AND/OR/NOT/allOf/anyOf can go when combining constraints. 

Yes, kinda along the lines of your EGFR and MET/TP53 examples. I was was also thinking, for instance, where a category can be satisfied in more than one way - e.g. where your criteria for amplification might be either an increased copy number or a duplication; or where your criteria for 'BRCA1 mutation' might be either a simple variant or a structural variant.

**Reply by DanielPuthawala** [reply]
Hey @rhdolin I'm still not quite following yet either.  What language/system/type system is the expression data type from?  Like in the SQL sense?

I think we should put our heads together in a call and brainstorm this.  Currently, some of that, like your example regarding amplification can already be handled as a recipe.   But it's also clear to me that some examples CAN'T be handled this way, like the MET/TP53 case, which seems like an irreducibly complex CatVar.  This makes me think that perhaps the route forward (though maybe for release later down the road) is to create a higher-order type over `CategoricalVariant`, which could be defined as being, in contrast to `CategoricalVariant` which supports a single array of `constraints', this higher-order typed class allows for a series of `constraints` arrays, together with the boolean logic between them, so that you could say something like 
```
Where:
 x, y, z :  CategoricalVariant
P : HOCategoricalVariant

As random made up example:
P = (x AND y) NOT z
```
meaning a complex CatVar that requires two separate sets of constraints to be satisfied, and also requires a third set to not be satisfied.  And example of this would be something like:

```
MET Amplification and TP53 Loss, but not MET Gain-of-Function
```

**Reply by DanielPuthawala** [reply]
I'mm make an issue to where we can follow up on this and think

**Reply by brendanreardon** [reply]
@DanielPuthawala created an Issue (https://github.com/ga4gh/cat-vrs/issues/92) for this a bit ago and I have quote-copied this exchange to there. Please continue the conversation in that Issue 🙇

**Comment by rhdolin** [comment]
### Categorical variant - question about the example

I'm unclear why 'NM_004958.4(MTOR):c.5992_5993del (p.Met1998fs)' is used as an example of a categorical variant. This doesn't look to be a categorical variant.

**Reply by DanielPuthawala** [reply]
It's a canonical allele, so in addition to the specific contextual variant in the label, it also contains as members congruent variants on other builds etc.  But more importantly, it's being represented categorically.  Even a singleton set CatVar is still a CatVar.

**Reply by rhdolin** [reply]
@DanielPuthawala I think it actually makes it harder to describe what a 'categorical variant' is, if a simple variant can be a categorical variant. In the readthedocs, we talk about NC_000007.13:g.140453136A>T vs. BRAF V600E, noting that the latter is a categorical variant. But based on the cited example, so could be the former (e.g. by submitting my assayed variant to ClinVar).

**Reply by DanielPuthawala** [reply]
Well, you're not wrong, but I think we need to be able to do that in order to do assayed-to-categorical variant matching.

As for the example, in the readthedocs,  its true that that's now a 100% accurate explanation, but I've yet to come up with abetter one that is as intuitive but captures the real distinction.   VRS handles contextual variants, ie specific variants in the context of some specific assay, genome, etc.  On the other hand, Cat-VRS  handles variants treated as set-theoretic objects, abstracted from the context in which they are assayed, and instead used as labels for classes tied to knowledge, and for which I think it's ultimately irrelevant whether the set contains as it's member a single contextual variant or a much more expansive set of contextual variant, and with other CatVars as definable subsets.

**Comment by rhdolin** [comment]
### Defining Location Constraint - question about the example

Example shows a location constraint with start and end **integers**, but should probably be start and end **ranges**.

**Reply by DanielPuthawala** [reply]
We're currently following VRS conventions, where locations typically contain start and end ranges, but for which unambiguous ranges can be replaced with an integer, so eg:

Location((4,4),(10,10)) = Location (4,10)

In the past I pushed for always using explicit ranges, and the group pushed back on the grounds that (1) this allows for shorter  variant entries, (2) that it's not unduly ambiguous, and (3) that GKS already has the tools to support parsing integers as unambiguous ranges, so it wasn't an implementation issue.

**Reply by rhdolin** [reply]
Thanks @DanielPuthawala , I see your point. But for CNVs, isn't it often the case that, say, inner endpoints are known whereas outer endpoints are not? If I'm interpreting the example correctly, it seems to suggest that we know the precise endpoints for this CNV.

**Reply by DanielPuthawala** [reply]
Ah, no, it can handle those cases, too. (this is all inherited from VRS)
So let's say we have a [range](https://github.com/ga4gh/vrs/blob/2a9e2ea7740747e1bcc4b556110fb92c72420b7c/schema/vrs/vrs-source.yaml#L475) with endpoints A and B.
A and B are each ranges (so the full range is actually a tuple of ranges):
```((A1, A2),(B1, B2))```
Where A1, A2, B1, and B2 can either be integers or a null value (here I'll use "?" just for readibility)

So a really specific could be: ((4,4),(10,10))
Which as I said above by convention is evaluated to mean the same thing as (4,10)

Here is a range where the outermost possible endpoints are known, but the innermost endpoints are not known: ((4,?),(?,10))
any subsequence including or between 4 and 10 are possible in this range, including 4-10, 4-5, 6-8, 7-7, etc.

Here is a range where the innermost endpoints are known, but the outermost possible endpoints are not known: ((?,4),(10,?))
Any sequence that covers at least positions 4-10 are possible given this range, including 4-10, 2-10, 4-15, 3-20, etc.

And here's a range that is maximally vague: ((?,?),(?,?)), which admits any sequence, regardless of length or position.

Does that make sense to you?

**Reply by rhdolin** [reply]
Thanks @DanielPuthawala. Yes, makes sense. My question really was about the specific example - I haven't seen CNVs with precise endpoints, so I would have expected the example to have ranges.

**Reply by brendanreardon** [reply]
This was represented in Issue https://github.com/ga4gh/cat-vrs/issues/87 and I have now quote copied this exchange as a comment on the Issue. Please continue the conversation there 🙇

**Comment by rhdolin** [comment]
### Protein Sequence Consequence - need to have a structured representation of the protein sequence 

Example should at lease include a valid pHGVS because as it stands, we only know that this example is for EGFR L858R via the label or alternativeLabels, both of which are text fields.

**Reply by afrubin** [reply]
The Protein Sequence Consequence includes a `DefiningAlleleConstraint` that has a VRS allele, but I agree that including a pHGVS somewhere would make sense. Maybe under the `extensions` section where the CIViC entries are?

Would the pHGVS be a property of the allele object specified in `vrs.json` per the example?

**Reply by rhdolin** [reply]
Thanks @afrubin , I missed that. Is there some reason why we can't include a canonical HGVS or SPDI as part of the DefiningAlleleConstraint so that one doesn't have to figure out how to unpack the VRS identifier?

**Reply by brendanreardon** [reply]
Thank you both! I totally agree, having HGVS representations within the examples should make the implementation more tractable. 

Last week Daniel [created an issue, #87](https://github.com/ga4gh/cat-vrs/issues/87) about updating the examples and I made a note to add HGVS representations for all examples too. Daniel brought up that the HGVS representations would likely fall under [mappings](https://cat-vrs.readthedocs.io/en/latest/concepts/imported/ConceptMapping.html#conceptmapping), since it is a representation in another system, but extension would work, too.

**Reply by afrubin** [reply]
I see the point, but we should definitely only support VRS for the `DefiningAlleleConstraint` for consistency. Keeping this object minimal is valuable so it's easy to pull out and validate, as adding extra fields that may be in some other format would complicate variant matching and validation. Adding it to `mappings` makes perfect sense to me.

**Reply by ahwagner** [reply]
I agree that we should only support VRS Alleles for the defining allele constraint, and will also point out that VRS Alleles have several things that can make downstream implementation easier, including a dedicated datatype ([expressions](https://vrs.ga4gh.org/en/2.0.0-ballot.2024-11/concepts/AdditionalDataTypes/index.html#expression)) for other syntax representations that can be added to any variation object under the `expressions` slot.

For example, [here](https://github.com/ga4gh/vrs/blob/2.x/examples/SPDI_expansion.yaml) is a VRS variant with a SPDI expression.

**Reply by afrubin** [reply]
It seems to me like we have multiple good options to solve this problem and store an HGVS representation:
* `mappings` in the CatVRS object 
* `expressions` in the VRS object linked to the `DefiningAlleleConstraint`

Of these two I favor the latter, because the HGVS representation is a property of the variant not the category. 

If we think about the Protein Sequence Consequence use case, the advantage is even more clear. We wouldn't want a mix of protein and nucleotide HGVS strings in the `mappings`, and downstream users are very likely to want the HGVS for the category members in addition to the `DefiningAlleleConstraint` variant.

Having said that, I see that in the example there are `mappings` to specific nucleotide variants (e.g. [CA126713](https://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_canonicalid?canonicalid=CA126713)) as `relatedMatch` objects. Is this going to scale well? Is the intention for the implementation to pull the `expressions` out of the `members` to generate a set of `relatedMatch` `mappings`?

Are there use cases I'm not thinking of where we would want this in `mappings` instead, or is this an issue that could be solved by clearly documenting that the intended solution is to use the VRS `expressions` slot?

Maybe my question is really about the purpose of the `mappings` field in general, and its intended use.

**Reply by DanielPuthawala** [reply]
As you say, Alan, `expressions` within the VRS objects is the most precise place to stick  the matched HGVS representation.  There are several places where an HGVS representation _can go_ in Cat-VRS, but it's not usually a very good fit, computationally, because whereas HGVS is representing contextual variants, Cat-VRS is representing categorical variants, ie sets of contextual variants.  As @rhdolin has mentioned elsewhere, this distinction is muddied a bit because some CatVars are singleton sets: they are set objects with one and only one contextual variant member.  This makes those CatVars synonymous with the VRS entries, but still distinct for our use-cases for search and matching.

So with that said, where could HGVS representations live in Cat-VRS?  One common place will probably be in `label`, which can be whatever people want to call it, and will probably be especially useful for variant classes where the 'canonical' name in the literature happens to be an HGVS string.  For singleton CatVars, `mappings` also works, as you can make an `exact match`, or for larger but for CatVars that are contextually well-clustered (like BRAF V600E, because exist in a relatively small range of genomic positions), a `mapping` to HGVS with a `related match` _may_ be useful. (and also might be not be, I don't claim to know).  And finally, within the `expressions` that appear with the VRS objects in `members`. This is, admittedly, the same as saying that HGVS just belongs in VRS, but does mean that Cat-VRS does have a place for such entities that can pull their HGVS `expressions` along with them.  However, the `members` field is not required to be exhaustive, so the abilit yot match based on HGVS will depend on whether or not the corresponding VRS object happens to be in that field or not, which to my mind poses a similar problem for search by HGVS string matching as does having `related matches` in `mappings`.   

But of course, that's not how we designed Cat-VRS to be enable search.  Instead, the way to be able to find what your looking for is to compute the singleton CatVar representation of your HGVS variant in Cat-VRS, and then query your CatVars by `constraints`, which is precisely how `recipes` work.

So in sum, there's a couple places where HGVS expressions could lvie in Cat-VRS, but I think outside of singleton CatVars, the best place is in the members, but also that the use of HGVS for matching in Cat-VRS is probably limited at best.

**Reply by afrubin** [reply]
Thanks @DanielPuthawala for the detailed response. It's good to see that Cat-VRS has the flexibility to handle all of these cases, and I'm sure that the preferred ways of handling each situation will make their way into the documentation as they are discovered through use.
